pkg/cmd/image/prune.go

```diff
@@ -0,0 +1,99 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package image
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+// Prune will remove all dangling images. If all is specified, will also remove all images not referenced by any container.
+func Prune(ctx context.Context, client *containerd.Client, options types.ImagePruneOptions) error {
+	var (
+		imageStore     = client.ImageService()
+		contentStore   = client.ContentStore()
+		containerStore = client.ContainerService()
+	)
+	delOpts := []images.DeleteOpt{images.SynchronousDelete()}
+	removedImages := make(map[string][]digest.Digest)
+
+	imageList, err := imageStore.List(ctx)
+	if err != nil {
+		return err
+	}
+
+	var filteredImages []images.Image
+
+	if options.All {
+		containerList, err := containerStore.List(ctx)
+		if err != nil {
+			return err
+		}
+		usedImages := make(map[string]struct{})
+		for _, container := range containerList {
+			usedImages[container.Image] = struct{}{}
+		}
+
+		for _, image := range imageList {
+			if _, ok := usedImages[image.Name]; ok {
+				continue
+			}
+
+			filteredImages = append(filteredImages, image)
+		}
+	} else {
+		for _, image := range imageList {
+			_, tag := imgutil.ParseRepoTag(image.Name)
+
+			if tag == "" {
+				filteredImages = append(filteredImages, image)
+			}
+		}
```

Comment:
I think you can reuse this (`filteredImages = filterDangling(imageList, true)`:

https://github.com/containerd/nerdctl/blob/67c2e11a12262d796da8336fb300f31a9adb6b9a/pkg/cmd/image/list.go#L126-L139

Reply:
I see it. My logic was copied from there.

Reply:
May need to move to imgutil.